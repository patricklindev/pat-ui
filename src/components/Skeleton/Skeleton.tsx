import * as React from 'react';
import { classNames } from '../../utils/classNames';

// Animation types for the skeleton component.
export type SkeletonAnimation = 'pulse'
    | 'wave'
    | false;

// Skeleton variant types. 
export type SkeletonVariant = 'circular'
    | 'rectangular'
    | 'rounded'
    | 'text'
    | string;

export interface ISkeletonProps {
    /** The animation type. If false, the animation effect is disabled. */
    animation?: SkeletonAnimation;
    /** The component used for the root node. Either a string to use a HTML element or a component. */
    component?: string | React.ElementType;
    /**  Height of the skeleton. Useful when you don't want to adapt the skeleton to a text element but for instance a card. */
    height?: number | string;
    /** The type of content that will be rendered. */
    variant?: SkeletonVariant;
    /** Width of the skeleton. Useful when the skeleton is inside an inline element with no width of its own. */
    width?: number | string;
}

export type PatSkeletonProps = React.PropsWithChildren<ISkeletonProps> & React.HTMLAttributes<React.ElementType>;

export const Skeleton: React.FC<PatSkeletonProps> = (props) => {
    const {
        animation,
        component: Component = 'div',
        height,
        variant,
        width,
        children,
        style,
        ...otherProps
    } = props;

    const classes = classNames('skeleton', {
        [variant as string]: !!variant,
        [animation as string]: !!animation
    });

    const styles: React.CSSProperties = {
        ...style,
        ...(width !== undefined && { width }),
        ...(height !== undefined && { height }),
    };

    return (
        <Component className={classes} style={styles} {...otherProps}>
            {children}
        </Component>
    );
};

export default Skeleton;
