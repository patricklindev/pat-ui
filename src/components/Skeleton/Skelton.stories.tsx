import React from 'react';
import Skeleton from './Skeleton';

export default {
    title: 'Skeleton',
    component: Skeleton,
};

export const DefaultSkeleton = () => (
    <Skeleton width='15rem' height='1rem' />
);

export const PulseSkeleton = () => (
    <Skeleton width='15rem' height='1rem' animation='pulse' />
);

export const WaveSkeleton = () => (
    <Skeleton width='15rem' height='1rem' animation='wave' />
);