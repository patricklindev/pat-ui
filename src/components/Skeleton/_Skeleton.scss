.skeleton {
    background-color: rgb(216, 216, 216); // light gray background
    border-radius: 4px;
    overflow: hidden;

    &.pulse {
        animation: pulse 1.5s infinite ease-in-out;
    }

    &.wave {
        position: relative;
        overflow: hidden;

        &::after {
            content: "";
            display: block;
            position: absolute;
            width: 50%;
            height: 100%;
            background: linear-gradient(to right, transparent, rgba(255, 255, 255, .2), transparent);
            animation: wave 1.5s infinite;
        }
    }

    &.circular {
        border-radius: 50%;
    }

    &.rectangular {
        border-radius: 0;
    }

    &.rounded {
        border-radius: 20px;
    }

    &.text {
        height: 16px;
        width: 100%;
        border-radius: 4px;
    }
}

@keyframes pulse {
    0% {
        opacity: 1;
    }

    50% {
        opacity: 0.4;
    }

    100% {
        opacity: 1;
    }
}

@keyframes wave {
    0% {
        transform: translateX(-100%);
    }

    50% {
        transform: translateX(-50%);
    }

    100% {
        transform: translateX(100%);
    }
}