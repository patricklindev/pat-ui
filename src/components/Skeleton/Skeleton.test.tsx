import { render } from '@testing-library/react';
import Skeleton from './Skeleton';
import React from 'react';

describe('Skeleton', () => {
    // snapshot testing
    it('should match snapshot', () => {
        // const { asFragment } = render(<Progress pgValue={0}></Progress>);
        // expect(asFragment()).toMatchSnapshot();
        const { container } = render(<Skeleton></Skeleton>);
        expect(container.firstChild).toMatchSnapshot();
    });
})