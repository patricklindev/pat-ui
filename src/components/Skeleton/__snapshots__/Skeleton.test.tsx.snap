// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Skeleton should match snapshot 1`] = `
<div
  class=" skeleton text"
/>
`;
